File: tests/fixtures/sample.php
<?php

namespace App\Billing;

interface PaymentProcessor
{
    public function process(int $amount): TransactionId;
}

trait Retryable
{
    public function withRetry(callable $operation, int $attempts = 3)
    {
        return $operation();
    }
}

// Implements PaymentProcessor against the StripeClient SDK.
class StripePaymentProcessor implements PaymentProcessor
{
    use Retryable;

    private string $apiKey;

    public function __construct(string $apiKey)
    {
        $this->apiKey = $apiKey;
    }

    public function process(int $amount): TransactionId
    {
        return $this->withRetry(fn() => $this->charge($amount));
    }

    private function charge(int $amount): TransactionId
    {
        return TransactionId::of($amount);
    }
}

File: tests/fixtures/sample.rb
module Billing
  # Mixed into every processor that talks to PaymentGateway.
  module Retryable
    def with_retry(attempts = 3)
      yield
    end
  end

  class PaymentProcessor
    def process(amount)
      raise NotImplementedError, "PaymentProcessor is abstract"
    end
  end

  class StripePaymentProcessor < PaymentProcessor
    include Retryable

    def initialize(api_key)
      @api_key = api_key
    end

    def process(amount)
      with_retry { charge(amount) }
    end

    private

    def charge(amount)
      TransactionId.of(amount)
    end
  end
end
